restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java

```diff
@@ -187,25 +188,33 @@ public QueryExecutor(DataStore dataStore) {
   }
 
   public Flowable<ResultSet> execute(BoundQuery query, int pageSize, ByteBuffer pagingState) {
-    // An empty paging state means the query was exhaused during previous execution
+    // An empty paging state means the query was exhausted during previous execution
     if (pagingState != null && pagingState.remaining() == 0) {
       return Flowable.empty();
     }
 
-    return fetchPage(query, pageSize, pagingState)
+    AtomicInteger pageNumber = new AtomicInteger();
+    pageNumber.set(1);
+
+    return fetchPage(query, 0, pageSize, pagingState)
         .compose( // Expand BREADTH_FIRST to reduce the number of "proactive" page requests
             FlowableTransformers.expand(
-                rs -> fetchNext(rs, pageSize, query), ExpandStrategy.BREADTH_FIRST, 1));
+                rs -> fetchNext(rs, pageNumber.getAndAdd(1), pageSize, query),
+                ExpandStrategy.BREADTH_FIRST,
+                1));
   }
 
-  private Flowable<ResultSet> fetchPage(BoundQuery query, int pageSize, ByteBuffer pagingState) {
+  private Flowable<ResultSet> fetchPage(
+      BoundQuery query, int pageNumber, int basePageSize, ByteBuffer pagingState) {
+    int pageSize = Math.max((int) Math.pow(2, pageNumber) * basePageSize, 10);
+    final int pageSizeFinal = Math.min(pageSize, 10000);
```

Comment:
What's the significance of `10000` here?

Reply:
A comment here would be good I agree -- my first thought was that one would usually cap with `max` (no higher than 10,000) but here it's at least 10k. Which may make sense but would be good to know the logic.